Rework transport tag allocation logic

joushou opened the pull request:
New tag allocation logic that avoids tag collisions, and errors out if a tag cannot be allocated.

stevvooe commented:
cc @justincormack

stevvooe commented on transport.go:
```diff
@@ -150,13 +150,31 @@ func (t *transport) handle() {
-			// about outstanding requests.
-			tags++
-			fcall := newFcall(tags, req.message)
+			var selected Tag
```
Do you mind compartmentalizing this into a method?

joushou replied:
Done!

stevvooe commented on transport.go:
```diff
@@ -150,13 +150,31 @@ func (t *transport) handle() {
+			var selected Tag
+			if req.message.Type() == Tversion {
+				// Tversion can only use NOTAG.
+				if _, exists := outstanding[NOTAG]; exists {
```
Please double check, but I believe that we never have version messages when there is an active transport. Please see https://github.com/docker/go-p9p/blob/master/version.go#L18. When the client is created, we do a [synchronous negotiation](https://github.com/docker/go-p9p/blob/master/client.go#L19), then the transport is created.

All we need to do here is ensure that `NOTAG` is not allocated and is it is encountered, there is a protocol error.

joushou replied:
I simply performed the check to guarantee correct tag allocation. I do not know what your plans are with regards to the use of transport.

I can remove the check if you want. I have no opinion on the matter.

stevvooe commented on transport.go:
```diff
@@ -150,13 +150,31 @@ func (t *transport) handle() {
+				}
+
+				exists := true
+				for selected = 0; exists && selected < NOTAG; selected++ {
```
This does a linear search through the tag space, making tag allocation more expensive as there are more outstanding messages. If we keep track of `selected`, this can be avoided. Collisions can be managed with linear probing (odd, I know, but this is what is happening), where if a collision is found, we go to the next entry. 

Consider a slice. The maximum bound for memory usage per session for the tag, aside from hash overhead (fill factor `λ`, where `λ` < 1) and the value side of the hash, will be ~`(N(size(p)+size(tag))/λ`, which is to 640K on 64-bit with `λ = 1`, and larger with a more realistic fill factor. With a slice, the memory becomes bounded by `N*size(p)`, which is just 512KB on 64-bit. We will also benefit in look since the values are linearly arranged in memory, which is optimal with a linear probing strategy. The trade off here is that holes cost us, but holes exist only as long as outstanding messages still exist. The holes will be proportional in size to the age of the oldest outstanding message.

joushou replied:
I have reintroduced the previous selection hint. I used it in my own implementation as well, but removed it as a simplification, thinking it could somewhat be considered a premature optimisation.

Having 65535 pending requests for a single client is unrealistic. A more realistic number of alive tags would be, idk, 10 to 100 tags, in which case we get <=1KB with `λ = 1`, but still 512KB for an array. Remember that tags are client local.

I agree that iterating on a map is a terrible idea, but I was weighing it against a permanent array that would be orders of magnitude larger. Even with thousand tags open, the overhead of using a permanent array is quite significant.

stevvooe replied:
> Even with thousand tags open, the overhead of using a permanent array is quite significant.

Go has a feature called slices which can avoid a full allocation of the tag space. It becomes a little more tricky that a map when handling wraparound, but we the memory usage is well-bounded in the common case vs the map, which can allocate a lot of extra memory.

Albeit, I'd be satisfied if we could simply avoid the linear scan through the maps keyspace. A hint plus linear probe should be sufficient.

joushou replied:
I know what a slice is. I said array intentionally. :)

In the simple mode, the slice cap would be equal to or greater than the highest tag in use. The slice cannot shrink without manually replacing it with a new slice with copied content, whereas the map can do whatever it wants (Not sure what it does in this scenario, actually). Furthermore, the slice will over-grow (512KB + 25% worst-case, IIRC) if we append.

We could make a `tagpool` type to take care of all this logic (check for slot, grow if none found and len < NOTAG, search from end to first tag to slice and shrink if len is much shorter than cap and cap is >100), but... That's a benchmark and an optimisation for another day. I prefer to have easily reproducible numbers when making performance claims and modifications.

stevvooe replied:
@joushou Sorry, wasn't sure what you meant there.

Agreed. It will likely be quite hard to match the efficiency of map's allocation algorithm. Sounds like we should avoid this without careful benchmarks.

joushou commented:
Summary of changes:
1. Tag allocation logic extracted to a method.
2. Tag hint to avoid unnecessary iteration (re)introduced.
3. newFcall tag reset logic removed. (It's not convenient, and blurs the lines of tag responsibility).

stevvooe commented on fcall.go:
```diff
@@ -107,11 +107,6 @@ type Fcall struct {
 }
 
 func newFcall(tag Tag, msg Message) *Fcall {
-	switch msg.Type() {
```
Double check that this doesn't break negotiation.

joushou replied:
clientnegotiate and servernegotiate both hardcode NOTAG to the call. Negotiation to my rather pedantic server also works with 9pr.

stevvooe commented on transport.go:
```diff
@@ -95,6 +96,43 @@ func (t *transport) send(ctx context.Context, msg Message) (Message, error) {
+	}
+
+	// Look for the first tag that doesn't exist in the map and return it.
+	for i := 0; i < 0xFFFF; i++ {
```
This doesn't handle wraparound. Double check the modulo arithmetic here, but I think this is sufficient: `for h := hint+1; h != hint; h++`.

stevvooe replied:
O nm, you handle this below. (Doh!)

stevvooe commented:
@joushou One check on the tag for negiotiation process and we're good!

joushou commented:
I have removed that check, and added commentary that the map must not contain NOTAG (it will otherwise give an erroneous tag pool depletion error when there is still an available tag).

stevvooe commented:
LGTM

@justincormack PTAL

justincormack commented:
LGTM

stevvooe commented:
@joushou Thanks for the great contribution!

Before merging, we'll need you to [sign your commits](https://github.com/docker/go-p9p/blob/master/CONTRIBUTING.md).

(Sorry, normally, this is automated, but we have yet to set this up for go-p9p)

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "tagpool" git@github.com:joushou/go-p9p.git somewhere
$ cd somewhere
$ git rebase -i HEAD~4
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

joushou commented:
All 4 commits signed.

stevvooe commented:
@joushou Cheers!
